### DISCUSSION TITLE
This app is AMAZING, but won’t run on my raspberry pi

### DISCUSSION THREAD

**Discussion by agoldh20** [jdholtz/auto-southwest-check-in#116] [General]
Hello!

I travel often on Sundays with Southwest and have used this repo for my last 4 trips. I noticed when my MacBook goes to sleep though the checkin doesn’t work, so I put it on my raspberry pi figuring that never goes to sleep.

However my pi is headless and I get webdriver issues when I try to use this. Any advice to get it to work with my pi? 

Keep up the great work!

**Comment by jdholtz** [comment]
Thanks! 

An ARM docker image can be downloaded [here](https://hub.docker.com/r/jdholtz/auto-southwest-check-in) for your Pi. You can see if the webdriver works when using that (if you aren’t already using it).

**Comment by jdholtz** [comment]
Also, what is the error you get from your webdriver?

**Reply by agoldh20** [reply]
```Process Process-1:
Traceback (most recent call last):
  File "/home/adam/auto-southwest-check-in/lib/webdriver.py", line 164, in _init_driver
    driver = Chrome(
  File "/home/adam/.local/lib/python3.9/site-packages/seleniumwire/undetected_chromedriver/webdriver.py", line 61, in __init__
    super().__init__(*args, **kwargs)
  File "/home/adam/.local/lib/python3.9/site-packages/undetected_chromedriver/__init__.py", line 429, in __init__
    super(Chrome, self).__init__(
  File "/home/adam/.local/lib/python3.9/site-packages/selenium/webdriver/chrome/webdriver.py", line 84, in __init__
    super().__init__(
  File "/home/adam/.local/lib/python3.9/site-packages/selenium/webdriver/chromium/webdriver.py", line 101, in __init__
    self.service.start()
  File "/home/adam/.local/lib/python3.9/site-packages/selenium/webdriver/common/service.py", line 96, in start
    self._start_process(self._path)
  File "/home/adam/.local/lib/python3.9/site-packages/selenium/webdriver/common/service.py", line 197, in _start_process
    self.process = subprocess.Popen(
  File "/usr/lib/python3.9/subprocess.py", line 951, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib/python3.9/subprocess.py", line 1823, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
OSError: [Errno 8] Exec format error: '/home/adam/.local/share/undetected_chromedriver/6817133d459cf7a8_chromedriver'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/lib/python3.9/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/usr/lib/python3.9/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/home/adam/auto-southwest-check-in/lib/flight_retriever.py", line 38, in monitor_flights
    self._schedule_reservations(flights)
  File "/home/adam/auto-southwest-check-in/lib/flight_retriever.py", line 66, in _schedule_reservations
    self.checkin_scheduler.schedule(confirmation_numbers)
  File "/home/adam/auto-southwest-check-in/lib/checkin_scheduler.py", line 36, in schedule
    self.refresh_headers()
  File "/home/adam/auto-southwest-check-in/lib/checkin_scheduler.py", line 48, in refresh_headers
    webdriver.set_headers()
  File "/home/adam/auto-southwest-check-in/lib/webdriver.py", line 58, in set_headers
    driver = self._get_driver()
  File "/home/adam/auto-southwest-check-in/lib/webdriver.py", line 139, in _get_driver
    driver = self._init_driver()
  File "/home/adam/auto-southwest-check-in/lib/webdriver.py", line 181, in _init_driver
    raise RuntimeError(
RuntimeError: Failed to initialize the webdriver after 3 attempts
```

**Reply by jdholtz** [reply]
This is a problem with `undetected_chromedriver` because it is downloading the chromedriver for the wrong architecture. The Docker image should fix that. 

If you do want to try to fix this yourself, you can download the chromedriver manually and patch it (directions can be found [here](https://github.com/ultrafunkamsterdam/undetected-chromedriver/issues/917#issuecomment-1327199992)). Then, you can specify the path to the chromedriver in the `config.json`.

**Reply by Hawkinzw** [reply]
Great tool! I got it working on an rPI 5 for a very similar reason. In case someone else runs into this I mostly took the advise above. Some more details:

Run: `sudo apt-get install chromium-chromedriver` to get the arm based chrome driver. Credit [this blog post](https://ivanderevianko.com/2020/01/selenium-chromedriver-for-raspberrypi).

Now you have a working chrome driver for arm, but selenium base will still try to go update to the latest from the web. I didn't follow the patch above. Instead I found it more straight forward to change line 136 in webdriver.py from `"mlatest" `to `"keep"` like it does for docker. That prevents selenium from trying to check/get the latest driver. 

My `config.json` has `"browser_path": "/usr/bin/chromium-browser"` as the last line to point to the driver.

**Comment by agoldh20** [comment]
hmmmmm
```
~ $ sudo docker pull jdholtz/auto-southwest-check-in
Using default tag: latest
latest: Pulling from jdholtz/auto-southwest-check-in
no matching manifest for linux/arm/v6 in the manifest list entries
```

I think I might have to try the manual chromedriver route

**Reply by jdholtz** [reply]
Unfortunately, there is only support for ARM64 as the Python libraries this script uses don’t build on ARMv6.

**Reply by agoldh20** [reply]
if thats the case, will the manual chrome driver route work?

**Reply by jdholtz** [reply]
I’m not sure, but it’s definitely worth trying.

**Reply by agoldh20** [reply]
what should the output be when running this via docker? I did a mock res number and name and it gave me a long random string back

**Reply by jdholtz** [reply]
You can check the script’s output by doing `docker logs <container-name>`. Additionally, you can run the Docker container without the `-d` flag to view the output directly. 

If you want more information, you can use the `-v` flag after the image name to make the script’s output verbose.

**Reply by agoldh20** [reply]
amazing! it works with docker on my pi! yaya thanks again for all the help!

**Reply by jdholtz** [reply]
Great! How did you download the docker image? Docker tried to download the ARMv6 and failed because it doesn’t exist.

**Reply by agoldh20** [reply]
I have a few pis around my house and one of them supports the 64 bit piOs. So I put docker on that one. I struggle with vim and nano so as soon as I figure out how to get IDEA to work via ssh I’ll dig into my pi zero

**Reply by jdholtz** [reply]
@agoldh20 have you been able to get the Docker image to build on ARMv6? It apparently is working for ARMv7 now, but when I tried to build ARMv6, it said there was no chromium package. If that is the case, I don't think ARMv6 support will ever come.

**Reply by agoldh20** [reply]
I'm pulling the latest image right now. my zero2 is very very laggy with docker so I'm trying on my 3

**Reply by agoldh20** [reply]
```
selenium.common.exceptions.WebDriverException: Message: unknown error: session deleted because of page crash
from unknown error: cannot determine loading status
from tab crashed
```
doesn't look like it

**Comment by camjones1708** [comment]
Was there a 32bit arm image every successfully created?

**Reply by camjones1708** [reply]
I am attempting to get it to run on my raspberry pi w/ chromium-chromedriver and am close but it's failing due to lack of cdp support I believe. If anyone has already walked this path, tips would be appreciated.

**Reply by jdholtz** [reply]
There is not, but if you are able to get it working without modifying the image (or making any changes not compatible with other architectures), I can build an official image for it. Unfortunately, I don’t have a 32bit ARM machine to test this on (also, the Docker image is currently having issues with logging in (429 status code)).

**Reply by camjones1708** [reply]
I _think_ it's due to cdp being incompatible in selenium with older versions of chrome. I can't find any newer 32bit versions of chrome, so I'll follow the other posts and switch to a newer one of my raspberry pi's.

**Reply by camjones1708** [reply]
I was able to get the docker image to run on my rpi 4, but its incredibly slow and fails to check in. Is there anything in the config that needs to be done to get it to run? 

Failed to check in to flight <CONF>for <FIRST> <LAST>. Reason: Too Many Requests 429.
Check in at this url: https://mobile.southwest.com/check-in

**Reply by jdholtz** [reply]
Currently no, as tracked in #201, there is no fix or mitigation for that issue

**Reply by camjones1708** [reply]
I see, I'll follow for information to the docker bug. I am still attempting to setup to run in console but am unable to instantiate the driver correctly on my arm64 raspberry pi. I am able to instantiate a driver with the Selenium WebDriver but not the driver.

**Reply by jdholtz** [reply]
Do you mean the driver is failing to be instantiated when run inside the script or when you try to run it manually?

**Reply by camjones1708** [reply]
I've been trying a bunch of different workarounds. Currently, I'm working on [undetected_chromedriver](https://github.com/ultrafunkamsterdam/undetected-chromedriver/issues/917), as with the uc_cdp_enabled flag, it leverages the undetected_chromedriver which doesn't seem to support arm64 architecture. 

So if I just use chromedriver, it works but then I can't leverage the listener you have set up. And undetected can't be executed or chrome won't start with it.

**Reply by camjones1708** [reply]
For anyone that finds this thread, I was able to get it to successfully run in console with the following driver/chomium versions installed and by installing box64.
https://stackoverflow.com/questions/76857893/is-there-a-known-working-configuration-for-using-selenium-on-linux-arm64

https://github.com/ptitSeb/box64/blob/main/docs/COMPILE.md
sudo wget https://ryanfortner.github.io/box64-debs/box64.list -O /etc/apt/sources.list.d/box64.list
wget -qO- https://ryanfortner.github.io/box64-debs/KEY.gpg | sudo gpg --dearmor -o /etc/apt/trusted.gpg.d/box64-debs-archive-keyring.gpg
sudo apt update && sudo apt install box64-arm64 -y

**Reply by agoldh20** [reply]
I couldn't get this to work anymore. I ended up just buying a cheap mac mini on ebay that needed a harddrive for less than a rpi4 am using that now for this project. I'll still mess around with this on my rpi zero2 though as I dig into docker a bit more
